aspnet/mvc/models/validation.rst

```diff
@@ -1,8 +1,230 @@
+
+.. literalinclude:: validation/sample/Movie.cs
+   :language: c#
+   :lines: 7-34
+   :dedent: 4
+
+Simply reading through the model reveals the rules about data for this app, making it easier to maintain the code. Below are several popular built-in data annotations:
+
+- ``[CreditCard]``: Specifies the property as a credit card type.
+- ``[Compare]``: Compares two properties in a model. It works with various types, such as int, string, or date. 
+- ``[EmailAddress]``: Specifies the property with an email format.
+- ``[Phone]``: Specifies the property as a telephone format.
+- ``[Range]``: Sets the minimum and maximum values for multiple types of properties.
+- ``[RegularExpression]``: Forces the data to match a regular expression.
+- ``[Required]``: Makes a property required.
+- ``[StringLength]``: The maximum length for a string field.
+- ``[Url]``: Specifies the property as a URL.
+
+MVC supports any attribute that derives from ``ValidationAttribute`` for validation purposes. Many useful validation attributes can be found in the `System.ComponentModel.DataAnnotations <https://msdn.microsoft.com/en-us/library/system.componentmodel.dataannotations(v=vs.110).aspx>`_.
+
+There may be intances where you need more features than built-in attributes provide. For those times, you can create custom validation attributes by deriving from ``ValidationAttribute`` or change your model to subclass ``IValidatableObject``.
+
+Model State
+-----------
+Model state represents validation errors that were submitted with HTML form values. You can then send back model state errors to the client. For example, if a movie is marked as a preorder, it should not already be released to the public. The view will display any errors discovered and added to ``ModelState`` as shown in the example below:
+
+.. literalinclude:: validation/sample/MoviesController.cs
+   :language: c#
+   :emphasize-lines: 44-49
+   :lines: 21-59
+   :dedent: 4   
+
+MVC will continue validating fields until reaches the maximum number of errors (200). You can configure this number by inserting the following code into the ``ConfigureServices`` method in the ``Startup.cs`` file:
+
+.. literalinclude:: validation/sample/Startup.cs
+   :language: c#
+   :lines: 5-14
+   :dedent: 4
+
+Manual validation
+-----------------
+After model binding and validation are complete, you may want to repeat parts of it. For example, a user may have entered text in a field expecting an integer, or you may need to compute a value for a model's property. You can then reset the invalid value to something acceptable. If you change the model like this, you'll want to run validation again. To run validation manually, call the ``TryValidateModel`` method, as shown here: 
+
+.. literalinclude:: validation/sample/MoviesController.cs
```

Comment:
Suggest focusing on the `if (preorder)` block rather than inline the entire controller class.